File: mobile/modules/tuya-pairing/android/src/main/java/expo/modules/tuyapairing/TuyaPairingModule.kt
package expo.modules.tuyapairing

import android.app.Application
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

import com.thingclips.smart.home.sdk.ThingHomeSdk
import com.thingclips.smart.android.user.api.ILoginCallback
import com.thingclips.smart.android.user.bean.User
import com.thingclips.smart.home.sdk.bean.HomeBean
import com.thingclips.smart.home.sdk.callback.IThingGetHomeListCallback
import com.thingclips.smart.home.sdk.callback.IThingHomeResultCallback
import com.thingclips.smart.sdk.api.IThingActivator
import com.thingclips.smart.sdk.api.IThingActivatorGetToken
import com.thingclips.smart.sdk.api.IThingSmartActivatorListener
import com.thingclips.smart.sdk.bean.DeviceBean
import com.thingclips.smart.sdk.builder.ThingActivatorBuilder
import com.thingclips.smart.sdk.enums.ActivatorModelEnum

// ─────────────────────────────────────────────────────────────────────────────
// Module natif d'appairage Tuya (modele C) — Android. Encapsule le device activator
// du SDK Tuya (Smart Life App SDK 7.5.x, namespace com.thingclips.smart). Enregistre
// sous « TuyaPairing », consomme par src/native/tuyaPairing.ts.
//
// NB API : les classes/methodes Tuya marquees « VALIDATION » sont a confirmer au 1er
// build EAS (le SDK 7.5.x "Thing" peut differer). La structure (callbacks, promises) est stable.
// ─────────────────────────────────────────────────────────────────────────────

class StartPairingParams : Record {
  @Field var mode: String = "EZ"
  @Field var ssid: String = ""
  @Field var password: String = ""
  @Field var token: String = ""
  @Field var timeoutSec: Double = 100.0
}

class TuyaPairingModule : Module() {

  private var activator: IThingActivator? = null

  override fun definition() = ModuleDefinition {
    Name("TuyaPairing")

    // Initialise le SDK Tuya avec l'AppKey/AppSecret Android (servis par le backend).
    AsyncFunction("init") { appKey: String, appSecret: String, promise: Promise ->
      val app = appContext.reactContext?.applicationContext as? Application
      if (app == null) {
        promise.reject("no_context", "Application context indisponible", null)
        return@AsyncFunction
      }
      ThingHomeSdk.init(app, appKey, appSecret)
      promise.resolve(null)
    }

    // Connecte le SDK au compte app Tuya de l'hote (provisionne par le PMS).
    // VALIDATION : loginWithUid pour un compte app cree via apps/{schema}/user — a aligner
    // avec username_type cote backend createAppUser.
    AsyncFunction("loginAppAccount") { countryCode: String, username: String, password: String, promise: Promise ->
      ThingHomeSdk.getUserInstance().loginWithUid(countryCode, username, password, object : ILoginCallback {
        override fun onSuccess(user: User) {
          promise.resolve(mapOf("uid" to user.uid))
        }
        override fun onError(code: String, error: String) {
          promise.reject("login_failed", error, null)
        }
      })
    }

    // Recupere un token d'appairage (lie a un home -> on resout/cree le home par defaut).
    AsyncFunction("getPairingToken") { promise: Promise ->
      resolveHomeId(
        onResolved = { homeId ->
          ThingHomeSdk.getActivatorInstance().getActivatorToken(homeId, object : IThingActivatorGetToken {
            override fun onSuccess(token: String) {
              promise.resolve(token)
            }
            override fun onFailure(code: String, error: String) {
              promise.reject("token_failed", error, null)
            }
          })
        },
        onError = { err -> promise.reject("home_error", err, null) }
      )
    }

    // Lance l'activator (provisioning Wi-Fi local) et resout avec le device appaire.
    AsyncFunction("startPairing") { params: StartPairingParams, promise: Promise ->
      startPairing(params, promise)
    }

    // Annule un appairage en cours.
    AsyncFunction("stopPairing") { promise: Promise ->
      cleanup()
      promise.resolve(null)
    }
  }

  // ── Helpers ────────────────────────────────────────────────────────────────

  /** Resout l'id du home courant ; en cree un par defaut s'il n'existe pas. */
  private fun resolveHomeId(onResolved: (Long) -> Unit, onError: (String) -> Unit) {
    ThingHomeSdk.getHomeManagerInstance().queryHomeList(object : IThingGetHomeListCallback {
      override fun onSuccess(homeBeans: MutableList<HomeBean>?) {
        val first = homeBeans?.firstOrNull()
        if (first != null) {
          onResolved(first.homeId)
        } else {
          ThingHomeSdk.getHomeManagerInstance().createHome(
            "Baitly", 0.0, 0.0, "", emptyList(),
            object : IThingHomeResultCallback {
              override fun onSuccess(bean: HomeBean) = onResolved(bean.homeId)
              override fun onError(errorCode: String, errorMsg: String) = onError(errorMsg)
            }
          )
        }
      }
      override fun onError(errorCode: String, error: String) = onError(error)
    })
  }

  private fun startPairing(params: StartPairingParams, promise: Promise) {
    // BLE non gere dans cette tranche (EZ/AP couvrent l'appairage Wi-Fi). BLE = enhancement.
    if (params.mode == "BLE") {
      promise.reject("ble_unsupported", "Mode BLE pas encore supporte (a venir). Utilisez EZ ou AP.", null)
      return
    }

    val context = appContext.reactContext?.applicationContext
    if (context == null) {
      promise.reject("no_context", "Application context indisponible", null)
      return
    }

    val model = if (params.mode == "AP") ActivatorModelEnum.THING_AP else ActivatorModelEnum.THING_EZ

    val builder = ThingActivatorBuilder()
      .setContext(context)
      .setSsid(params.ssid)
      .setPassword(params.password)
      .setActivatorModel(model)
      .setTimeOut(params.timeoutSec.toLong())
      .setToken(params.token)
      .setListener(object : IThingSmartActivatorListener {
        override fun onError(errorCode: String, errorMsg: String) {
          promise.reject("pairing_failed", errorMsg, null)
          cleanup()
        }

        override fun onActiveSuccess(devResp: DeviceBean) {
          val device = mapOf(
            "devId" to (devResp.devId ?: ""),
            "name" to (devResp.name ?: ""),
            "productId" to (devResp.productId ?: ""),
            "category" to (devResp.getCategory() ?: "")
          )
          promise.resolve(listOf(device))
          cleanup()
        }

        override fun onStep(step: String, data: Any?) {
          // Progression (facultatif) — pourra etre relaye via un event Expo plus tard.
        }
      })

    activator = ThingHomeSdk.getActivatorInstance().newActivator(builder)
    activator?.start()
  }

  private fun cleanup() {
    activator?.stop()
    activator?.onDestroy()
    activator = null
  }
}

File: mobile/modules/tuya-pairing/ios/TuyaPairingModule.swift
import ExpoModulesCore
import ThingSmartHomeKit

// ─────────────────────────────────────────────────────────────────────────────
// Module natif d'appairage Tuya (modele C) — encapsule le device activator du SDK
// Tuya (Smart Life App SDK 7.5.x). Enregistre sous le nom « TuyaPairing », consomme
// par la facade TS src/native/tuyaPairing.ts.
//
// Flux : init(appKey, appSecret) au demarrage -> loginAppAccount (compte app de l'hote
// fourni par le PMS) -> getPairingToken (lie a un home) -> startPairing (activator local
// EZ/AP) -> resout avec le(s) device(s) appaire(s).
//
// NB API : les selecteurs Tuya marques « VALIDATION » sont a confirmer au 1er build EAS
// (le SDK 7.5.x "Thing" peut differer legerement). La structure (delegate, promises) est stable.
// ─────────────────────────────────────────────────────────────────────────────

struct StartPairingParams: Record {
  @Field var mode: String = "EZ"
  @Field var ssid: String = ""
  @Field var password: String = ""
  @Field var token: String = ""
  @Field var timeoutSec: Double = 100
}

public class TuyaPairingModule: Module {

  private var activator: ThingSmartActivator?
  private var activatorDelegate: ActivatorDelegate?
  private var pairingPromise: Promise?
  private var pairedDevices: [[String: Any]] = []

  public func definition() -> ModuleDefinition {
    Name("TuyaPairing")

    // Initialise le SDK Tuya avec l'AppKey/AppSecret iOS (servis par le backend).
    AsyncFunction("init") { (appKey: String, appSecret: String, promise: Promise) in
      ThingSmartSDK.sharedInstance().start(withAppKey: appKey, secretKey: appSecret)
      promise.resolve(nil)
    }

    // Connecte le SDK au compte app Tuya de l'hote (provisionne par le PMS).
    // VALIDATION : selecteur de login par username (le compte est cree via apps/{schema}/user,
    // username_type a confirmer cote backend createAppUser). A aligner avec le vrai SDK.
    AsyncFunction("loginAppAccount") { (countryCode: String, username: String, password: String, promise: Promise) in
      ThingSmartUser.sharedInstance().login(byUserName: username,
                                            countryCode: countryCode,
                                            password: password) {
        let uid = ThingSmartUser.sharedInstance().uid ?? ""
        promise.resolve(["uid": uid])
      } failure: { error in
        promise.reject("login_failed", error?.localizedDescription ?? "Echec de connexion au compte Tuya")
      }
    }

    // Recupere un token d'appairage. Tuya lie le token a un home -> on resout/cree le home par defaut.
    AsyncFunction("getPairingToken") { (promise: Promise) in
      self.resolveHomeId { homeId, error in
        guard let homeId = homeId else {
          promise.reject("home_error", error ?? "Aucun home Tuya disponible")
          return
        }
        ThingSmartActivator.getTokenWithHomeId(homeId) { token in
          promise.resolve(token ?? "")
        } failure: { error in
          promise.reject("token_failed", error?.localizedDescription ?? "Echec d'obtention du token d'appairage")
        }
      }
    }

    // Lance l'activator (provisioning Wi-Fi local) et resout avec le(s) device(s) appaire(s).
    AsyncFunction("startPairing") { (params: StartPairingParams, promise: Promise) in
      self.startPairing(params: params, promise: promise)
    }

    // Annule un appairage en cours.
    AsyncFunction("stopPairing") { (promise: Promise) in
      self.cleanup()
      promise.resolve(nil)
    }
  }

  // ── Helpers ────────────────────────────────────────────────────────────────

  /// Resout l'id du home courant ; en cree un par defaut s'il n'existe pas.
  private func resolveHomeId(_ completion: @escaping (Int64?, String?) -> Void) {
    let manager = ThingSmartHomeManager()
    manager.getHomeList { homes in
      if let home = homes?.first {
        completion(home.homeId, nil)
      } else {
        // Pas de home -> on en cree un (requis par l'activator).
        manager.addHome(withName: "Baitly",
                        geoName: "",
                        rooms: [],
                        latitude: 0,
                        longitude: 0) { homeId in
          completion(homeId, nil)
        } failure: { error in
          completion(nil, error?.localizedDescription ?? "Echec de creation du home")
        }
      }
    } failure: { error in
      completion(nil, error?.localizedDescription ?? "Echec de lecture des homes")
    }
  }

  private func startPairing(params: StartPairingParams, promise: Promise) {
    // BLE non gere dans cette tranche (EZ/AP couvrent l'appairage Wi-Fi). BLE = enhancement.
    if params.mode == "BLE" {
      promise.reject("ble_unsupported", "Mode BLE pas encore supporte (a venir). Utilisez EZ ou AP.")
      return
    }

    self.pairedDevices = []
    self.pairingPromise = promise

    let mode: ThingActivatorMode = (params.mode == "AP") ? .AP : .EZ
    let delegate = ActivatorDelegate { [weak self] device, error in
      guard let self = self else { return }
      if let error = error {
        self.pairingPromise?.reject("pairing_failed", error.localizedDescription)
        self.cleanup()
        return
      }
      guard let device = device else { return }
      self.pairedDevices.append([
        "devId": device.devId ?? "",
        "name": device.name ?? "",
        "productId": device.productId ?? "",
        "category": device.category ?? ""
      ])
      // Un device recu = succes ; l'UI peut relancer pour appairer d'autres appareils.
      self.pairingPromise?.resolve(self.pairedDevices)
      self.cleanup()
    }

    let activator = ThingSmartActivator()
    activator.delegate = delegate
    self.activator = activator
    self.activatorDelegate = delegate

    activator.startConfigWiFi(mode,
                              ssid: params.ssid,
                              password: params.password,
                              token: params.token,
                              timeout: TimeInterval(params.timeoutSec))
  }

  private func cleanup() {
    self.activator?.stopConfigWiFi()
    self.activator?.delegate = nil
    self.activator = nil
    self.activatorDelegate = nil
    self.pairingPromise = nil
  }
}

// Delegate de l'activator Tuya : relaie chaque device recu (ou erreur) au closure.
final class ActivatorDelegate: NSObject, ThingSmartActivatorDelegate {
  private let onDevice: (ThingSmartDeviceModel?, Error?) -> Void

  init(onDevice: @escaping (ThingSmartDeviceModel?, Error?) -> Void) {
    self.onDevice = onDevice
  }

  func activator(_ activator: ThingSmartActivator!,
                 didReceiveDevice deviceModel: ThingSmartDeviceModel!,
                 error: Error!) {
    onDevice(deviceModel, error)
  }
}
